torch/utils/checkpoint.py

```diff
@@ -335,7 +339,275 @@ def forward(input):
+    def backward(ctx, *grad_outputs):
+        raise AssertionError("Did not expect to backward on this graph")
+
+class _CheckpointFrame():
+    def __init__(self, recompute_fn):
+        self.recompute_fn = recompute_fn
+        self.input_saver = None
+        self.weak_holders: List[ReferenceType] = []
+        # We store this as a weakkeydictionary so that in the case of a partial
+        # backward, the entries in the dict are cleared alongside the Holder
+        # which will be removed when the SavedVariable is cleared.
+        self.recomputed: DefaultDict[int, weakref.WeakKeyDictionary[_Handle, torch.Tensor]] = \
+            defaultdict(weakref.WeakKeyDictionary)
+        self.recomp_counter: DefaultDict[int, int] = defaultdict(int)
+        self.is_recomputed: DefaultDict[int, bool] = defaultdict(bool)
+
+# See Rule 5
+class _StopRecomputationError(Exception):
+    pass
+
+class _recomputation_hook(torch.autograd.graph.saved_tensors_hooks):
+    def __init__(self, target_frame_ref: ReferenceType, gid: int):
+        def pack_hook(x):
+            target_frame = target_frame_ref()
+            assert target_frame is not None
+            recomp_idx = target_frame.recomp_counter[gid]
+            target_frame.recomp_counter[gid] += 1
+
+            if recomp_idx >= len(target_frame.weak_holders):
+                # We run into this case when early stop is not enabled and do
+                # grad within checkpoint.
+                return x.detach()
+            holder = target_frame.weak_holders[recomp_idx]()
+
+            if holder is not None:
+                # See Rule 6: [ Multiple backwards ] above
+                if holder.handles.get(gid, None) is None:
+                    holder.handles[gid] = _Handle()
+                target_frame.recomputed[gid][holder.handles[gid]] = x.detach()
+
+            if _enable_checkpoint_early_stop and \
+               target_frame.recomp_counter[gid] == len(target_frame.weak_holders):
+                raise _StopRecomputationError()
+            # See Rule 6: [ Basic case ] above
+            return x.detach()
+
+        def unpack_hook(x):
+            return x
+
+        super().__init__(pack_hook, unpack_hook)
+
+class _checkpoint_hook(torch.autograd.graph.saved_tensors_hooks):
+    def __init__(self, frame):
+        def pack_hook(_unused_x):
+            # See Rule 4 above
+            holder = _Holder()
+            frame.weak_holders.append(weakref.ref(holder))
+            return holder
+
+        def unpack_hook(holder):
+            gid = torch._C._current_graph_task_id()
+            if gid == -1:
+                # generate a temporary id if we trigger unpack outside of a backward call
+                gid = int(uuid.uuid4())
+
+            if not frame.is_recomputed[gid]:
+                ctx = frame.input_saver.grad_fn
+                args = ctx.get_args(ctx.saved_tensors)
+
+                try:
+                    # pass gid in in case we do reentrant backward
```

Comment:
Oh one question I asked myself is if one could just query for gid again and get the same result, and I think the answer is that querying gid again gets the same result unless we have reentrant backward.

I will remove this though since from a perf perspective querying again is probably worse anyway, so we'd probably want to pass gid either way!